### PR TITLE
[3.11] Bug 1693025: Add CSRF to token request endpoint

### PR DESCRIPTION
**Pull request by stlaz**
This is a 3.11 backport of openshift/origin#22767 which adds CSRF validation to OAuth server token requests created from its web console.

@openshift/sig-auth

### CODE REVIEW

File: pkg/oauthserver/oauthserver/auth.go [hunk 1]
```diff
@@ -140,7 +140,7 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
 	)
 	server.Install(mux, urls.OpenShiftOAuthAPIPrefix)
 
-	tokenRequestEndpoints := tokenrequest.NewEndpoints(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient)
+	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient, c.ExtraOAuthConfig.OAuthAccessTokenClient)
```
Comment (enj) [inline pkg/oauthserver/oauthserver/auth.go]: No `OAuthAccessTokenClient`

File: pkg/oauthserver/server/grant/grant.go [hunk 2]
```diff
@@ -109,18 +108,16 @@ func (l *Grant) Install(mux oauthserver.Mux, paths ...string) {
 }
 
 func (l *Grant) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	headers.SetStandardHeaders(w)
-
-	user, ok, err := l.auth.AuthenticateRequest(req)
+	authResponse, ok, err := l.auth.AuthenticateRequest(req)
```
Comment (enj) [inline pkg/oauthserver/server/grant/grant.go]: This should still be `user, ok, err := l.auth.AuthenticateRequest(req)`

File: pkg/oauthserver/server/headers/oauthbasic.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package headers
```
Comment (enj) [inline pkg/oauthserver/server/headers/oauthbasic.go]: Undo this commit - the handler chain changes are too aggressive of a change to backport.

File: pkg/oauthserver/server/tokenrequest/tokenrequest.go [hunk 4]
```diff
@@ -19,81 +18,54 @@ import (
...
 	// osinOAuthClientGetter is used to initialize osinOAuthClient.
 	// Since it can return an error, it may be called multiple times.
 	osinOAuthClientGetter func() (*osincli.Client, error)
-	// ready is closed to signal that osinOAuthClient is no longer nil.
-	// Nothing sends on ready so <-ready only returns when it has been closed.
-	ready chan struct{}
-	// initLock guards reads and writes to osinOAuthClient when it could still be nil.
-	initLock sync.Mutex
+
+	// to check if we need the logout link for the bootstrap user
```
Comment (enj) [inline pkg/oauthserver/server/tokenrequest/tokenrequest.go]: You should not have this at all.

File: pkg/oauthserver/oauthserver/auth.go [hunk 5]
```diff
@@ -140,7 +140,7 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
 	)
 	server.Install(mux, urls.OpenShiftOAuthAPIPrefix)
 
-	tokenRequestEndpoints := tokenrequest.NewEndpoints(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient)
+	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient, c.ExtraOAuthConfig.OAuthAccessTokenClient, c.getCSRF())
```
Comment (enj) [inline pkg/oauthserver/oauthserver/auth.go]: Remove `OAuthAccessTokenClient`

File: pkg/oauthserver/oauthserver/oauth_apiserver.go [hunk 6]
```diff
@@ -207,20 +204,26 @@ func (c completedOAuthConfig) New(delegationTarget genericapiserver.DelegationTa
...
-	handler = genericfilters.WithCORS(handler, genericConfig.CorsAllowedOriginList, nil, nil, nil, "true")
-	handler = genericfilters.WithTimeoutForNonLongRunningRequests(handler, genericConfig.LongRunningFunc, genericConfig.RequestTimeout)
-	handler = genericapifilters.WithRequestInfo(handler, genericapiserver.NewRequestInfoResolver(genericConfig))
-	handler = genericfilters.WithPanicRecovery(handler)
+	// add back the Authorization header so that WithOAuth can use it even after WithAuthentication deletes it
+	// WithOAuth sees users' passwords and can mint tokens so this is not really an issue
+	handler = headers.WithRestoreAuthorizationHeader(handler)
+
+	// this is the normal kube handler chain
+	handler = genericapiserver.DefaultBuildHandlerChain(handler, genericConfig)
```
Comment (enj) [inline pkg/oauthserver/oauthserver/oauth_apiserver.go]: Undo (see other comment).

File: pkg/oauthserver/server/grant/grant.go [hunk 7]
```diff
@@ -99,28 +97,21 @@ func NewGrant(csrf csrf.CSRF, auth authenticator.Request, render FormRenderer, c
...
 	if err != nil || !ok {
 		l.redirect("You must reauthenticate before continuing", w, req)
 		return
 	}
 	switch req.Method {
 	case "GET":
-		l.handleForm(user, w, req)
+		l.handleForm(authResponse, w, req)
```
Comment (enj) [inline pkg/oauthserver/server/grant/grant.go]: Undo

File: pkg/oauthserver/server/grant/grant.go [hunk 8]
```diff
@@ -99,28 +97,21 @@ func NewGrant(csrf csrf.CSRF, auth authenticator.Request, render FormRenderer, c
...
 	case "POST":
-		l.handleGrant(user, w, req)
+		l.handleGrant(authResponse, w, req)
```
Comment (enj) [inline pkg/oauthserver/server/grant/grant.go]: Undo

File: pkg/oauthserver/server/tokenrequest/tokenrequest.go (hunk 4 above)
Comment (stlaz) [reply pkg/oauthserver/server/tokenrequest/tokenrequest.go]: Slipped past me, thanks.

**Comment by stlaz** [comment]
Hopefully addressed all the comments. I tried to modify the CSRF token in a resent POST request and the server failed properly with being unable to verify it.

**Comment by stlaz** [comment]
Fixed formatting so that `test/verify` passes...

**Comment by stlaz** [comment]
/retest
unit test failed to build coz it got SIGKILL

File: pkg/oauthserver/oauthserver/auth.go (hunk 1 above)
Comment (enj) [inline pkg/oauthserver/oauthserver/auth.go]: ffb4fa19c2f81e85e86f866603da103b69913e88 should not have `c.ExtraOAuthConfig.OAuthAccessTokenClient`

File: pkg/oauthserver/oauthserver/auth.go [hunk 9]
```diff
@@ -140,7 +140,7 @@ func (c *OAuthServerConfig) WithOAuth(handler http.Handler) (http.Handler, error
 	)
 	server.Install(mux, urls.OpenShiftOAuthAPIPrefix)
 
-	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient, c.ExtraOAuthConfig.OAuthAccessTokenClient)
+	tokenRequestEndpoints := tokenrequest.NewTokenRequest(c.ExtraOAuthConfig.Options.MasterPublicURL, c.getOsinOAuthClient)
```
Comment (enj) [inline pkg/oauthserver/oauthserver/auth.go]: And then 5dddf472704af722175fb10e6226f3d6345c181e would not be trying to remove it.

**Comment by stlaz** [comment]
Patchset cleanup done and the bounce castle is now slowly being inflated for the backport party.

**Comment by stlaz** [comment]
/hold
The registry client test fails consistently, needs investigation

**Comment by stlaz** [comment]
@openshift/sig-developer-experience you were enabling/disabling the test in master several times, what shall be done here? Note that the backport will go deep in versions.

**Comment by stlaz** [comment]
Disabled the failing test, I'll open BZ for the DevExp team once the PR merges.

**Comment by stlaz** [comment]
BZ tracking the disabled test: https://bugzilla.redhat.com/show_bug.cgi?id=1719156

**Pull request merged**